bookShelf/books/views.py

```diff
@@ -17,147 +19,435 @@
-	path = request.GET.get('path',"/")
-	print (path)
-	depth = int(request.GET.get('depth',3))
+	f = jsc.path_to_dict(DATABASE_DIR, DATABASE_DICT_FILE_NAME)
+	path = request.GET.get('path', "/")
+	depth = int(request.GET.get('depth', 3))
 	try:
-		db = jsc.navigate_path(f,path)
+		db = jsc.navigate_path(f, path)
 	except Exception as e:
 		print("invalid path")
 		return Response({})
 	else:
-		truncated_db = jsc.truncate_db(db,depth)
+		truncated_db = jsc.truncate_db(db, depth)
 		return Response(truncated_db)
-#testAPI
+
+
+@api_view()
+def APIsearch(request):
+	"""
+		gives a list of all path_prefix objects matching search terms
+	"""
+	f = jsc.path_to_dict(DATABASE_DIR, DATABASE_DICT_FILE_NAME)  ####can this be drier? repeated code
+	keyword_list = (request.GET.get('query', "")).split()
+	path = request.GET.get('path', "/")
+	path_prefix = search.get_path_prefix(path)
+	try:
+		db = jsc.navigate_path(f, path)
+	except Exception as e:
+		print("invalid path")
+		return Response({})
+	else:
+		result = search.search_dic(db, path_prefix, keyword_list)
+		return Response({"result": result})
+
+
+# testAPI
 @api_view()
 def heartbeat(request):
-    return Response({"message":"Server is up!"});
+	return Response({"message": "Server is up!"})
+
+
+def zip_courses():
+	"""
+		function to intelligently zip all the courses
+		depending upon if their zips have been deleted 
+		due to less number of downloads or not 
+	"""
+	for root, dirs, files in os.walk(DATABASE_DIR, topdown=True):
+		flag = 0
+		for name in dirs:
+			if len(name) == 2:
+				flag = 1
+				break
+		if root[len(DATABASE_DIR) + 1:].count(os.sep) == 0 and flag == 0:
+			for course in dirs:
+				course_path = os.path.join(root, course)
+				zip_file = os.path.join(root, course) + ".zip"
+				if os.path.exists(zip_file) and os.path.isfile(zip_file):
+					zf = zipfile.ZipFile(zip_file, "a")
+					for dirname, sudirs, files in os.walk(os.path.join(root, course)):
+						for filename in files:
+							if not filename.lower().endswith('.meta'):
+								path = os.path.join(dirname, filename)
+								to_write = os.path.relpath(path, course_path)
+								zf.write(path, arcname=to_write)
+					zf.close()
+				else:
+					is_changed = 0
+					for dirname, sudirs, files in os.walk(os.path.join(root, course)):
+						for filename in files:
+							if is_changed == 0:
+								zf = zipfile.ZipFile(zip_file, "w")
+								is_changed = 1
+							if not filename.lower().endswith('.meta'):
+								path = os.path.join(dirname, filename)
+								to_write = os.path.relpath(path, course_path)
+								zf.write(path, arcname=to_write)
+							else:
+								meta_file = os.path.split(filename)[1]
+								file_loc = get_file_loc(meta_file)[1]
+								path = os.path.join(dirname, filename)
+								to_write = os.path.join(os.path.split(os.path.relpath(path, course_path))[0],
+														os.path.split(file_loc)[1])
+								zf.write(file_loc, arcname=to_write)
```

Comment:
This causes file not found error, when clicking on finalize approvals, in a course that just got its first file.